contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,470 @@
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        // subtraction is safe because we checked positionLiquidity is gte liquidity
+        position.liquidity = uint128(positionLiquidity - liquidity);
+    }
+
+    // deploy ERC-20 contract
+    function beforeInitialize(address, IPoolManager.PoolKey calldata key, uint160) external override returns (bytes4) {
+        require(key.tickSpacing == 60, "Tick spacing must be default");
+
+        // deploy erc20 contract
+
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        BalanceDelta balanceDelta = hookModifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: -int256(int128(position.liquidity))
+            })
+        );
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+        uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtRatioAtTick(MIN_TICK),
+            TickMath.getSqrtRatioAtTick(MAX_TICK),
+            uint256(uint128(-balanceDelta.amount0())),
+            uint256(uint128(-balanceDelta.amount1()))
+        );
+
+        // reinvest everything
+        hookModifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: int256(int128(liquidity))
+            })
+        );
+
+        // update position
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        position.tokensOwed0 = 0;
+        position.tokensOwed1 = 0;
+    }
+
+    function beforeModifyPosition(
+        address sender,
+        IPoolManager.PoolKey calldata key,
+        IPoolManager.ModifyPositionParams calldata params
+    ) external override returns (bytes4) {
+        // check msg.sender
+        require(sender == address(this), "sender must be hook");
+
+        if (block.number > blockNumber) {
+            blockNumber = block.number;
+            HookPosition storage position = poolToHookPosition[key.toId()];
+
+            if (position.tokensOwed1 > 0 || position.tokensOwed0 > 0) {
+                _rebalance(key);
```

Comment:
my hot take is that you should rebase the erc20 instead of constantly rebasing the lp position. its far too expensive to rebalance it every single block. i think if you rebase the erc20 then apply the new position then there isnt any concerns (could be wrong).

some divergence will happen because of fee returns but if you rebase them on mint/burns it should be fine

Reply:
Yah, was talking to Hayden a bit and was thinking maybe it's not necessary to rebalance every block, (crucial on every burn tho). The one thing that rebasing the lp position does that the ERC20 doesn't, is that it auto-compounds the fees which was a really nice V2 benefit.

Reply:
I guess rebasing on every burn will just means the LP will get less benefit from auto compounding. like if they mint a position, and then burn before any other LP burns their position, then they will not get any autocompounding from their position, whereas having it done every block kind of guarantees that you will get the autocompounding benefits as long as u keep you position in for more than a block, but maybe it is not worth the gas cost